src/zkbob/utils/CustomABIDecoder.sol

```diff
@@ -190,4 +190,38 @@ contract CustomABIDecoder {
     function _memo_permit_holder() internal pure returns (address r) {
         r = address(uint160(_loaduint256(memo_permit_holder_pos + memo_permit_holder_size - uint256_size)));
     }
+
+    function _mpc_signatures_pos() internal pure returns ( uint256 pos) {
+        uint256 t = _tx_type(); 
+        if (t==3 || t == 0) {
+          pos = _sign_r_vs_pos() + sign_r_vs_size;
+        } else {
+            pos = _sign_r_vs_pos();
+        }
+    }
+    
+
+    function _mpc_message() internal pure returns (bytes calldata message) {
+        uint256 message_length = _mpc_signatures_pos();
```

Comment:
If we compose MPC message from the data on the contract (`roots[_pool_index]`) + call data (`_transfer_pub()`, `_transfer_proof()`, `_tree_proof()` etc), the contract can be 100% sure that data used to produce signatures is the same data as was used to build zkproof.

Reply:
d96c0092f48d3f29ecf83ab1f6445471d3d1428c

- added current root and transfer root to the challenge in regular transactions
- added current root to the challenge in direct deposits

Reply:
It's not critical, but we also use `_pool_id()` as public input, taking it not from the calldata but directly from the contract. If we aim to ensure the complete integrity of public inputs, we need to include the pool ID as well.

Reply:
fixed in [d96c009](https://github.com/zkBob/zkbob-contracts/pull/99/commits/d96c0092f48d3f29ecf83ab1f6445471d3d1428c)